File: lc_soln/1448_count_good_nodes_in_binary_tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        stack = [(root, root.val)]
        result = 0
        while stack:
            node, max_val = stack.pop()
            max_val = max(max_val, node.val)
            if node.val >= max_val:
                result += 1
            if node.left is not None:
                stack.append((node.left, max_val))
            if node.right is not None:
                stack.append((node.right, max_val))
        return result


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/1448_count_good_nodes_in_binary_tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        int result = 0;
        std::stack<TreeNode*> node_stack;
        std::stack<int> val_stack;
        node_stack.push(root);
        val_stack.push(root->val);
        while (!node_stack.empty()) {
            TreeNode* node = node_stack.top();
            int max_val = std::max(val_stack.top(), node->val);
            node_stack.pop();
            val_stack.pop();
            if (node->val >= max_val) {
                result++;
            }
            if (node->left != nullptr) {
                node_stack.push(node->left);
                val_stack.push(max_val);
            }
            if (node->right != nullptr) {
                node_stack.push(node->right);
                val_stack.push(max_val);
            }
        }
        return result;
    }
};

int main() {
    return 0;
}
